### PR TITLE
Add support for JSON Feed Version 1

### PR DESCRIPTION
**Pull request by tantalic**
This adds support for feeds in the new JSON Feed Version 1 format (specification: https://jsonfeed.org/version/1). 

Changes and feedback are welcome!

### CODE REVIEW

**Comment by tantalic** [comment]
The PR currently fails to build on go 1.6 or earlier as it is using the `SetIndent` method on `json,Encoder` to set indentation parameters which match those which were used for XML encoding for RSS and Atom feeds. This method was added in go 1.7: https://golang.org/doc/go1.7#encoding_json

I see to possible paths forward:

1. Change support to 1.7 and greater for this library. As version 1.7 was released almost 10 months ago at this point and is the oldest version currently receiving security releases this seems like a reasonable decision.
2. Change the PR to remove indentation from JSON. This would have the benefit of a slight reduction in JSON size while sacrificing readability and consistency with how JSON is formatted.

**Comment by kisielk** [comment]
I'm fine supporting only go1.7 and later going forward. I'll review this PR later tonight or this coming week. If you don't hear from me feel free to ping here.

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,6 @@
 language: go
 sudo: false
 go:
-  - 1.3
-  - 1.4
-  - 1.5
   - 1.7
   - 1.8
   - tip
```
Comment (elithrar) [inline .travis.yml]: While we're here:

```
matrix:
  include:
    - go: 1.7
    - go: 1.8
    - go: 1.x
    - go: master
  allow_failures:
    - go: master
```

`1.x` will always point to the latest version. `master` points to the latest rev, which is useful for identifying breaking changes in upcoming versions.

File: README.md [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 
-feeds is a web feed generator library for generating RSS and Atom feeds from Go
+feeds is a web feed generator library for generating RSS, Atom and JSON feeds from Go
 applications.
 
 ### Goals
 
  * Provide a simple interface to create both Atom & RSS 2.0 feeds
- * Full support for Atom and RSS2.0 spec elements
+ * Full support for Atom, RSS2.0, and JSON Feed Version 1 spec elements
```
Comment (elithrar) [inline README.md]: Have the `JSON Feed` text link to https://jsonfeed.org/ - being a very new spec, package users are less likely to be familiar with it.

File: feed.go [hunk 3]
```diff
@@ -105,3 +106,19 @@ func (f *Feed) ToRss() (string, error) {
 func (f *Feed) WriteRss(w io.Writer) error {
 	return WriteXML(&Rss{f}, w)
 }
+
+// creates an JSON representation of this feed
```
Comment (elithrar) [inline feed.go]: `ToJSON creates a JSON Feed representation of this feed.`

File: feed.go [hunk 4]
```diff
@@ -105,3 +106,19 @@ func (f *Feed) ToRss() (string, error) {
...
+func (f *Feed) ToJSON() (string, error) {
+	j := &JSON{f}
+	return j.ToJSON()
+}
+
+// Writes an JSON representation of this feed to the writer.
```
Comment (elithrar) [inline feed.go]: `WriteJSON writes a JSON Feed representation of this feed to the given writer`.

(if you can update the existing docstrings to follow this format, I'd be grateful!)

File: json.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		raw.EmbeddedJSONAttachment.Duration = time.Duration(nsec)
+	}
+
+	return nil
+}
+
+// JSONItem represents a single entry/post for the feed.
+type JSONItem struct {
+	Id            string           `json:"id"`
+	Url           string           `json:"url,omitempty"`
```
Comment (elithrar) [inline json.go]: `URL`

File: json.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	ExternalUrl   string           `json:"external_url,omitempty"`
```
Comment (elithrar) [inline json.go]: `ExternalURL`

File: json.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	Url  string `json:"url"`
+}
+
+// JSONFeed represents a syndication feed in the JSON Feed Version 1 format.
+// Matching the specification found here: https://jsonfeed.org/version/1.
+type JSONFeed struct {
+	Version     string      `json:"version"`
+	Title       string      `json:"title"`
+	HomePageUrl string      `json:"home_page_url,omitempty"`
+	FeedUrl     string      `json:"feed_url,omitempty"`
```
Comment (elithrar) [inline json.go]: `FeedURL`

File: json.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	Type string `json:"type"`
+	Url  string `json:"url"`
+}
+
+// JSONFeed represents a syndication feed in the JSON Feed Version 1 format.
+// Matching the specification found here: https://jsonfeed.org/version/1.
+type JSONFeed struct {
+	Version     string      `json:"version"`
+	Title       string      `json:"title"`
+	HomePageUrl string      `json:"home_page_url,omitempty"`
```
Comment (elithrar) [inline json.go]: `HomePageURL`

File: json.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	NextUrl     string      `json:"next_url,omitempty"`
+	Icon        string      `json:"icon,omitempty"`
+	Favicon     string      `json:"favicon,omitempty"`
+	Author      *JSONAuthor `json:"author,omitempty"`
+	Expired     *bool       `json:"expired,omitempty"`
+	Hubs        []*JSONItem `json:"hubs,omitempty"`
+	Items       []*JSONItem `json:"items,omitempty"`
+}
+
+type JSON struct {
```
Comment (elithrar) [inline json.go]: Exported type needs comment

**Comment by tantalic** [comment]
@elithrar - Most existing identifierers in this package that contain initialisms do not currently not following the Go conventions around constant casing (for example: `Url`,`Rss`, `Guid`, `Uri`, `Id`, `Href`, etc). 

This raises a couple of questions:

1. Is it better to stay consistent within the package or to adopt the generally accepted naming conventions for additions? 
2. If it is preferable to use `URL` should we also use `ID`? 
3. If it is preferable to stay consistent within the package should we use `Json` instead of `JSON`? (This was the instance where I did break from the package conventions. Somehow at the time it felt justifiably different from the other cases although now that I review it I'm not sure that is really the case given that the package uses `Rss`, `Atom` and `Xml` throughout.)

File: feed.go (hunk 4 above)
Comment (tantalic) [inline feed.go]: @elithrar I would be happy to go through and update/add some docstrings for the package. I will put those in separate PR in the future.

**Comment by kisielk** [comment]
I think you should stick to the conventions used in the package. I know `URL` is used elsewhere, but that ship has sailed as far as this package goes, so it's better to stick with `Url` unless we make a v2 and update everything.

File: json.go (hunk 9 above)
Comment (tantalic) [inline json.go]: @elithrar - As the `Atom` and `Rss` types do not have doccomments I am at a bit of a loss as to how to label this. Do you have a suggestion?

File: json.go (hunk 5 above)
Comment (tantalic) [inline json.go]: Not changed based on comment @kisielk about remaining internally consistent with other names in this package.

File: json.go (hunk 6 above)
Comment (tantalic) [inline json.go]: Not changed based on comment @kisielk about remaining internally consistent with other names in this package.

File: json.go (hunk 8 above)
Comment (tantalic) [inline json.go]: Not changed based on comment @kisielk about remaining internally consistent with other names in this package.

File: json.go (hunk 7 above)
Comment (tantalic) [inline json.go]: Not changed based on comment @kisielk about remaining internally consistent with other names in this package.

**Comment by tantalic** [comment]
@kisielk this should be ready for review.

**Comment by kisielk** [comment]
I think it looks good to me. Any other comments @elithrar  ?

**Comment by elithrar** [comment]
Thanks @tantalic!

**Pull request merged**